primitives/src/merkle_tree/merkle_tree_impl.rs

```diff
@@ -0,0 +1,413 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+use super::{
+    AppendableMerkleTree, ElementType, Hasher, LookupResult, MerkleTree, UpdatableMerkleTree,
+};
+use crate::errors::PrimitivesError;
+use ark_ff::Field;
+use ark_std::{borrow::Borrow, boxed::Box, marker::PhantomData, string::ToString, vec, vec::Vec};
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+use typenum::Unsigned;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct MerkleTreeImpl<E, H, LeafArity, TreeArity, F>
+where
+    LeafArity: Unsigned,
+    TreeArity: Unsigned,
+    F: Field,
+{
+    root: Box<MerkleNode<E, F>>,
+    height: usize,
+    capacity: u64,
+    num_leaves: u64,
+
+    _phantom_h: PhantomData<H>,
+    _phantom_la: PhantomData<LeafArity>,
+    _phantom_ta: PhantomData<TreeArity>,
+}
+
+impl<E, H, LeafArity, TreeArity, F> MerkleTree<F> for MerkleTreeImpl<E, H, LeafArity, TreeArity, F>
+where
+    E: ElementType<F>,
+    H: Hasher<F>,
+    LeafArity: Unsigned,
+    TreeArity: Unsigned,
+    F: Field,
+{
+    type ElementType = E;
+    type Hasher = H;
+    type LeafArity = LeafArity;
+    type TreeArity = TreeArity;
+    type Proof = MerkleProof<E, F>;
+    type BatchProof = MerkleNode<E, F>;
```

Comment:
Is it better to define a struct `MerkleBatchProof<E,F>` rather than using `MerkleNode`?

Reply:
I haven't implement the batch proof part. This is just a placeholder. 
Naturally for a batch proof, it contains a sub-tree. (Single proof is just a tree path, we can collect it as a vector).
I'm still deciding on the right representation for a sub-tree.